### PR TITLE
Add command line Brave Ads permission rule

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/23336

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [x] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Confirm ads are not served on production if `--enable-features`, `--force-fieldtrials`, `--variations-server-url`, `--variations-insecure-server-url`, `--fake-variations-channel`, `--variations-override-country`, `--force-fieldtrial-param` and `field-trial-handle` are used on the command-line in production environments

### CODE REVIEW

**Review by kylehickinson — APPROVED** [review]
iOS++

**Review by emerick — APPROVED** [review]
LGTM

**Review by zenparsing — APPROVED** [review]

File: components/brave_ads/browser/ads_service_impl.cc [hunk 1]
```diff
@@ -839,7 +840,55 @@ void AdsServiceImpl::DetectUncertainFuture(const uint32_t number_of_start) {
...
+  // to components which will then provide access to |kFeatureName| and
+  // command-line switches rather than using hard coded strings below.
+  const auto* command_line = base::CommandLine::ForCurrentProcess();
+  if (command_line->HasSwitch("variations-server-url") ||
+      command_line->HasSwitch("variations-insecure-server-url") ||
+      command_line->HasSwitch("fake-variations-channel") ||
+      command_line->HasSwitch("variations-override-country")) {
+    sys_info->did_override_command_line_args_flag = true;
+  } else {
+    std::string switch_value;
```
Comment (aseren) [inline components/brave_ads/browser/ads_service_impl.cc:854]: Could we collect switch values into base::flat_set?

File: components/brave_ads/browser/ads_service_impl.cc [hunk 2]
```diff
@@ -839,7 +840,55 @@ void AdsServiceImpl::DetectUncertainFuture(const uint32_t number_of_start) {
...
+    if (command_line->HasSwitch("force-fieldtrial-params")) {
+      switch_value +=
+          command_line->GetSwitchValueASCII("force-fieldtrial-params");
+    }
+    if (command_line->HasSwitch(switches::kFieldTrialHandle)) {
+      switch_value +=
+          command_line->GetSwitchValueASCII(switches::kFieldTrialHandle);
+    }
+
+    const std::vector<std::string> feature_names = {"AdRewards",
```
Comment (aseren) [inline components/brave_ads/browser/ads_service_impl.cc:868]: we could have something like this: https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ash/file_manager/file_tasks.cc;l=1030?q=kHtmlExtensions&ss=chromium

**Review by aseren — APPROVED** [review]
LGTM with couple of nits

File: components/brave_ads/browser/ads_service_impl.cc (hunk 1 above)
Comment (tmancey) [reply components/brave_ads/browser/ads_service_impl.cc:855]: Will fix this as part of decoupling

Comment (tmancey) [reply components/brave_ads/browser/ads_service_impl.cc:855]: Fixed

**Pull request merged**